File: PostOrderIter.scala
/**
 * Given a binary tree, provide a post order iterator.  The expected
 * solution is one that provides a stateful class and is O(n) over the tree.
 * 
 * I also provide a standard post order recursive solution that uses
 * continuations to avoid stack overflow.
 */

// Covariant tree model.
sealed trait Tree[+T]
case object Leaf extends Tree[Nothing]
case class Node[+T](left: Tree[T], value: T, right: Tree[T]) extends Tree[T]
object Node {
  def apply[T](value: T): Node[T] = Node(Leaf, value, Leaf)
}

/***
 * Stateful post order binary tree iterator using lambdas.
 */
class PostOrderIter[T](t: Tree[T]) {
  var ops: List[() => T] = List()
  pushNode(t)

  def hasNext(): Boolean = !ops.isEmpty

  // Pop an op.
  def next(): T = {
    val op = ops.head
    ops = ops.tail
    op()
  }

  private def pushNode: Tree[T] => Unit = {
    case Node(l, v, r) =>
      ops = (() => v) :: ops
      ops = (() => opNode(r)) :: ops
      ops = (() => opNode(l)) :: ops
    case Leaf => ()
  }

  private def opNode(t: Tree[T]): T = {
    pushNode(t)
    next()
  }
}

/**
 * This is a standard post order iteration of a tree using continuations.
 * NOTE: This is my preferred solution to this problem.
 */
def iterate[T](
  f: T => Unit,
  n: Tree[T],
  k: => Unit = {}
) {
  n match {
    case Leaf => k
    case Node(l, v, r) =>
      iterate(f, l, {
        iterate(f, r, {
          f(v)
          k
        })
      })
  }
}

def doTree(t: Tree[Int]) {
  val iter = new PostOrderIter(t)
  while (iter.hasNext()) {
    print(iter.next())
  }
  print(' ')
  iterate(print, t)
  println()
}

doTree(Leaf)
doTree(Node(1))
doTree(Node(Node(1), 2, Leaf))
doTree(Node(Leaf, 1, Node(2)))
doTree(Node(Node(1), 2, Node(3)))
doTree(Node(Node(Node(1), 2, Node(3)), 5, Node(Node(6), 7, Node(8))))

File: PostOrderIter.fs
(* This is another implementation of a stateful post order binary tree iterator.
   In this one, we build a stack of lambdas where each lambda represents the next
   operation, and returns the next value.  See `PushNode` for the core functionality.
*)
type Tree<'a> = Empty | Node of Tree<'a> * 'a * Tree<'a>

type PostOrderIter<'a>(root: Tree<'a>) as this =
    let mutable ops : List<unit -> 'a> = []
    do this.PushNode root

    member this.HasNext: bool =
        not <| List.isEmpty ops

    member this.Next(): 'a =
        match ops with
        | op :: opsTail ->
            ops <- opsTail
            op()
        | _ -> failwith "no more ops"

    member private this.OpNode (t: Tree<'a>) =
        this.PushNode t
        this.Next()

    member private this.PushNode(t: Tree<'a>) =
        match t with
        | Node(l, v, r) ->
            ops <- (fun () -> v) :: ops
            ops <- (fun () -> this.OpNode r) :: ops
            ops <- (fun () -> this.OpNode l) :: ops
        | Empty -> ()

let DoTree (t: Tree<'a>) =
    let iter = new PostOrderIter<'a>(t)
    while iter.HasNext do
        printf "%A " <| iter.Next()
    printfn ""

let Singleton(v: 'a) =
    Node(Empty, v, Empty)

let Node3(l: 'a, v: 'a, r: 'a) =
    Node(Singleton(l), v, Singleton(r))

[<EntryPoint>]
let main argv = 
    DoTree Empty
    DoTree <| Singleton 1
    DoTree <| Node3(1, 2, 3)
    DoTree <| Node(Node3(1, 2, 3), 4, Node3(5, 6, 7))
    
    0 // return an integer exit code
